### PR TITLE
Feature EF Core for repositories

### PR DESCRIPTION
**Pull request by nikcio**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

Superset of #14563

### What

This PR does the needed work to be able to migrate the Umbraco repositories from using NPOCO to EF Core under the hood. The following has been done (In addition to the changes in #14563):
- `DbContext` has been moved to its own folder
- Adds `UmbracoDatabaseFactory` for the EF Core implementation
- `IUmbracoDatabaseContract` has been added to have a framework agnosic way of interacting with EF Core
- `UmbracoDbContextBase` has been introduced to allow `IUmbracoDatabaseContract` not to depend on EF Core
- `UmbracoDbContext` now implements `UmbracoDbContextBase`
- `UmbracoDbContext` now overrides all `DbSet` and only has a `get;` available (This is best practice.)
- `DatabaseSchemaCreator.ValidateSchema` is now public to allow using it from the EF Core project.
- `IUmbracoDatabase` is now also `IUmbracoDatabaseContract`.
- NPOCO specific methods on `IUmbracoDatabase` has been marked `Obsolete`
- NPOCO specific methods on `IUmbracoDatabaseFactory` has been marked `Obsolete`
- `CreateDatabaseAsync` has been added to `IUmbracoDatabase` to allow async database creation
- Moved the old `UmbracoDatabaseFactory` service call (`AddSingleton`) to the EF Core project to make sure injection works for all non-converted services.
- Moved EF Core models* to `src/Umbraco.Cms.Infrastructure.Persistence/Models` (The models have no dependency on EF Core)
- The new `UmbracoDatabaseFactory` and `UmbracoDatabase` implementations now work as a proxy on methods that are obsolete. - Using the legacy implementation for these calls.
- Adds more checks when setting up DbContext
- Adds MoveToInfrastructure migration to update the modelsnapshot
- The test `Lock_Exceeds_Command_Timeout` tests specific implementation logic for NPOCO and has been changed to use the `LegacyDatabase` property on the new database model.

*The models for OpenIddict have been kept in the EF Core project because these models depend on EF Core directly. This shouldn't matter because the models are not used directly by the Umbraco code and therefore, don't need to be available on the `IUmbracoDatabase` implementation.

### Why

To allow Umbraco to move to EF Core as its ORM this step is needed to provide APIs that can be used in the Umbraco repositories while not breaking the current implementation of the Umbraco database abstraction for NPOCO.

### Notes

The integration test project uses the `Configure` method to set the connection string in the database factory. This will need to be changed so that it uses the `IOptions` so it doesn't create special cases where only the legacy database factory can be used because the connection string isn't set on the DbContext which can only be done when creating the services through the `IOptions`.

### Test

All Unit/Integration/E2E tests should pass.

Migration commands

_Prerequisites: Configure the 'src/Umbraco.Web.UI/appsettings.json' to have the correct connection string for the provider under test._

```sh
dotnet ef migrations add test -s src/Umbraco.Web.UI -p src/Umbraco.Cms.Persistence.EFCore.SqlServer -c UmbracoDbContext  -- --provider SqlServer
```
Creates an empty migration named test for SqlServer (Shouldn't affect the modelsnapshot either)

```sh
dotnet ef migrations add test -s src/Umbraco.Web.UI -p src/Umbraco.Cms.Persistence.EFCore.Sqlite -c UmbracoDbContext  -- --provider Sqlite
```
Creates an empty migration named test for Sqlite(Shouldn't affect the modelsnapshot either)

### People

@nikolajlauridsen @bergmania

### CODE REVIEW

**Comment by nikcio** [comment]
@nul800sebastiaan both #14563 and #14564 are mostly in response to my comment on https://github.com/umbraco/Umbraco-CMS/pull/14449#issuecomment-1637044620 it's meant to show the developers at HQ how (I believe) is the best way to implement EF Core into Umbraco in place of NPOCO. So I hope you can use it in some way even if that is to copy only some of the changes to another branch and merge it or to do your own twist on the changes completely disconnected from these changes I'm okay with that.

With that said I surely hope the PR can be merged in some way but that is completely up to you at HQ. I know that it's a ton of code changes and can be quite a mouthful to take in. But let me know if you have any questions about either my comments, #14563 or #14564 .

TLDR: This is mostly for demonstration purposes.

**Comment by nikcio** [comment]
# I took a swing at a POC for using EF Core in the Umbraco repositories - Here's what I encountered

### [[View file changes for POC](https://github.com/nikcio/Umbraco-CMS/compare/compare/branch/feature/ef-core-for-repositories-updated-from-contrib...nikcio:Umbraco-CMS:feature/poc/document-repository-ef)]

So I took some time to try using EF Core in a Umbraco repository to see how difficult it would be to use it in the current repositories. The method I tried using it on is `src/Umbraco.Infrastructure/Persistence/Repositories/Implement/DocumentRepository.cs` - `protected override void PersistNewItem(IContent entity)` called by `ContentService.Save()`. The specific implementation isn't interesting in this journey because it would require more work to have it working exactly right. This implementation isn't done and doesn't act correctly in all use cases. This can be seen in the failing tests in the branch 😅 

## Well enough intro let's see what I encountered:
- When trying to convert the method I noticed that the user ids being used in the tests are invalid and therefore I encountered that when trying to test the integration tests EF Core would throw a Foreign Key error. I've opened a PR to fix this here: #14639 
- We will need extensions for the `SaveChanges()` method to save the changes on the entities - This has been added to the `IUmbracoDatabase` interface and the EF Core implementation
- The current implementation for setting the connection string/Provider on the DbContext doesn't update when the connection string is changed. This will happen on a Umbraco install and also happens in every integration test as stated in the "Notes" section of this PR:
  - > Notes
The integration test project uses the Configure method to set the connection string in the database factory. This will need to be changed so that it uses the IOptions so it doesn't create special cases where only the legacy database factory can be used because the connection string isn't set on the DbContext which can only be done when creating the services through the IOptions.
  - This can be fixed by setting the connection string/provider when creating the DbContext and disposing every already created DbContext when a connection string is changed to ensure that every DbContext instance has the same connection. See the changes in the `src/Umbraco.Cms.Persistence.EFCore/DbContexts/UmbracoDbContext.cs`. This also means that the initial setups in the `src/Umbraco.Cms.Persistence.EFCore/Extensions/UmbracoEFCoreServiceCollectionExtensions.cs` aren't needed and can be removed.
- **[Important]** When writing the queries via Linq as you do in EF Core you quickly encounter that we're missing a lot of the extensions needed to do these operations. To fix this I created every extension that EF Core offers on `IQueryable` and a `DbSet` by copying the interface from the EF Core. See `src/Umbraco.Infrastructure/Extensions/IQueryableExtensions.cs`. These extensions then call a class I've called `src/Umbraco.Infrastructure/QueryBuilders/IEFCoreQueryBuilder.cs` which is implemented in the EF Core project, see `src/Umbraco.Cms.Persistence.EFCore/QueryBuilders/EFCoreQueryBuilder.cs`, and acts as the middleman between the Infrastructure project and the EF Core project keeping the projects decoupled. This class is provided statically and is added as a singleton because it doesn't include any state itself see: `src/Umbraco.Cms.Persistence.EFCore/Composition/UmbracoEFCoreComposer.cs`. Other models that are copied from the EF Core source code to bridge the gap are:
  - `src/Umbraco.Infrastructure/QueryBuilders/IQueryableInterfaces/IIncludableQueryable.cs`
  - `src/Umbraco.Infrastructure/QueryBuilders/QueryTrackingBehavior.cs`

## Implementation notes

Just to clarify again the implementation I worked on to see how this would work isn't complete and doesn't pass every test. I also tried refactoring the implementation to be easier to understand but the points above are the most valuable.

**Comment by nikcio** [comment]
I've come across a better way to go about this and will be creating a new POC PR and a discussion where I can better describe the intent. Therefore will this PR be deconstructed in a similar faction to #14563 

[When it's fully deconstructed I will close the PR in favor of the new ones 😄]

Status:

Targeted by #14639
- When trying to convert the method I noticed that the user ids being used in the tests are invalid and therefore I encountered that when trying to test the integration tests EF Core would throw a Foreign Key error.

Targeted by #14674 
- The current implementation for setting the connection string/Provider on the DbContext doesn't update when the connection string is changed. This will happen on a Umbraco install and also happens in every integration test as stated in the "Notes" section of this PR
- Adds more checks when setting up DbContext

Targeted by POC - TODO
- `DbContext` has been moved to its own folder
- Moved EF Core models* to `src/Umbraco.Cms.Infrastructure.Persistence/Models` (The models have no dependency on EF Core)

Will not be included in the new POC
- We will need extensions for the `SaveChanges()` method to save the changes on the entities - This has been added to the `IUmbracoDatabase` interface and the EF Core implementation
- Adds `UmbracoDatabaseFactory` for the EF Core implementation
- `IUmbracoDatabaseContract` has been added to have a framework agnosic way of interacting with EF Core
- `UmbracoDbContextBase` has been introduced to allow `IUmbracoDatabaseContract` not to depend on EF Core
- `UmbracoDbContext` now implements `UmbracoDbContextBase`
- `UmbracoDbContext` now overrides all `DbSet` and only has a `get;` available (This is best practice.)
- `DatabaseSchemaCreator.ValidateSchema` is now public to allow using it from the EF Core project. (This will be done differently)
- `IUmbracoDatabase` is now also `IUmbracoDatabaseContract`.
- NPOCO specific methods on `IUmbracoDatabase` has been marked `Obsolete`
- NPOCO specific methods on `IUmbracoDatabaseFactory` has been marked `Obsolete`
- `CreateDatabaseAsync` has been added to `IUmbracoDatabase` to allow async database creation
- Moved the old `UmbracoDatabaseFactory` service call (`AddSingleton`) to the EF Core project to make sure injection works for all non-converted services.
- The new `UmbracoDatabaseFactory` and `UmbracoDatabase` implementations now work as a proxy on methods that are obsolete. - Using the legacy implementation for these calls.
- Adds MoveToInfrastructure migration to update the modelsnapshot
- The test `Lock_Exceeds_Command_Timeout` tests specific implementation logic for NPOCO and has been changed to use the `LegacyDatabase` property on the new database model.

**Comment by nul800sebastiaan** [comment]
Hi again @nikcio - I appreciate you really wanting to work on this but as I said before, we don't want to waste your time! 

I've talked to some people at HQ and there's already a branch being worked on for this, more EF Core support in the core of Umbraco. I don't know the details, but please before you continue, talk to @Zeegaan and see if you two can team up instead. 

Regrettably I'll close this PR for now and thanks for all the work on it. I can only recommend for large PRs that you get in touch with us first (https://github.com/umbraco/Umbraco-CMS/discussions) and get a nod of approval of your plans otherwise we'll get into this situation again. 

Hopefully @Zeegaan can point you to his branch and at the very least you can look at his code and do a review on it, or collaborate where possible.

**Comment by nul800sebastiaan** [comment]
Ps. He's on holiday this week, so don't expect a response until next week.

**Comment by nikcio** [comment]
@nul800sebastiaan All good 👍
